BootloaderCorePkg/Library/PciEnumerationLib/PciEnumerationLib.c

```diff
@@ -1291,17 +1334,128 @@ PciScanRootBridges (
+
+  @retval EFI_SUCCESS             Create HOBs successfully
+  @retval EFI_INVALID_PARAMETER   Invalid parameters passing
+  @retval EFI_LOAD_ERROR          Invalid data integrity
+
+ **/
+EFI_STATUS
+BuildPciRootBridgeInfoHob (
+  IN      PCI_IO_DEVICE             *Root,
+  IN OUT  PCI_ROOT_BRIDGE_INFO_HOB  *RootBridgeInfoHob,
+  IN      UINT8                      MaxRootBridgeCount
+  )
+{
+  UINT32          Address;
+  UINT8           Index;
+  UINT8           Count;
+  PCI_BAR_TYPE    BarType;
+
+  if ((RootBridgeInfoHob == NULL) || (Root == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Count = RootBridgeInfoHob->Count;
+  if (Count >= MaxRootBridgeCount) {
+    return EFI_LOAD_ERROR;
+  }
+
+  Address = Root->Address;
+  RootBridgeInfoHob->Entry[Count].BusBase  = ((Address >> 8) & 0xFF);
+  RootBridgeInfoHob->Entry[Count].BusLimit = (UINT8)(Address & 0xFF);
+  for (Index = 0; Index < PCI_MAX_BAR; Index++) {
+    if (Root->PciBar[Index].Length == 0) {
+      continue;
+    }
+    BarType = Root->PciBar[Index].BarType;
+    if ((BarType == PciBarTypeUnknown) || (BarType > PciBarTypePMem64)) {
+      continue;
+    }
+    RootBridgeInfoHob->Entry[Count].Resource[BarType - 1].ResBase   = Root->PciBar[Index].BaseAddress;
+    RootBridgeInfoHob->Entry[Count].Resource[BarType - 1].ResLength = Root->PciBar[Index].Length;
+  }
+  RootBridgeInfoHob->Count++;
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Build PCI related HOBs
+
+  @param [in] RootBridge        A pointer of Root Bridges List
+  @param [in] RootBridgeCount   The number of found root bridges
+
+  @retval EFI_SUCCESS           Create HOBs successfully
+  @retval EFI_OUT_OF_RESOURCES  Out of Hob resource
+  @retval EFI_LOAD_ERROR        Invalid data integrity
+
+ **/
+EFI_STATUS
+BuildPciInfoHob (
+  IN  PCI_IO_DEVICE     *RootBridge,
+  IN  UINT8              RootBridgeCount
+  )
+{
+  EFI_STATUS                 Status;
+  LIST_ENTRY                *CurrentLink;
+  PCI_ROOT_BRIDGE_INFO_HOB  *RootBridgeInfoHob;
+  PCI_IO_DEVICE             *Root;
+  UINTN                      Length;
+
+  Status = EFI_SUCCESS;
+
+  Length  = sizeof (PCI_ROOT_BRIDGE_INFO_HOB);
+  Length += sizeof (PCI_ROOT_BRIDGE_ENTRY) * RootBridgeCount;
+  RootBridgeInfoHob = BuildGuidHob (&gLoaderPciRootBridgeInfoGuid, Length);
+  if (RootBridgeInfoHob == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  ZeroMem (RootBridgeInfoHob, Length);
+  RootBridgeInfoHob->Revision = 1;
+  RootBridgeInfoHob->Count    = 0;
+
+  CurrentLink = RootBridge->ChildList.ForwardLink;
+  while ((CurrentLink != NULL) && (CurrentLink != &RootBridge->ChildList)) {
+    Root = PCI_IO_DEVICE_FROM_LINK (CurrentLink);
+    Status = BuildPciRootBridgeInfoHob (Root, RootBridgeInfoHob, RootBridgeCount);
+    if (EFI_ERROR (Status)) {
+      break;
+    }
+
+    CurrentLink = CurrentLink->ForwardLink;
+  }
+
+  if (RootBridgeInfoHob->Count != RootBridgeCount) {
+    Status = EFI_LOAD_ERROR;
+  }
+
+  if (EFI_ERROR (Status)) {

```

Comment:
Maybe rollback Count to 0 is good enough?   EX:  You might still want the Revision info.

Reply:
Updated as commented. Thanks.